File: store_app/lib/core/models/store.dart
class StoreLocation {
  final double lat;
  final double lng;
  final String? geohash;

  const StoreLocation({required this.lat, required this.lng, this.geohash});

  factory StoreLocation.fromJson(Map<String, dynamic> j) => StoreLocation(
        lat: (j['lat'] as num).toDouble(),
        lng: (j['lng'] as num).toDouble(),
        geohash: j['geohash'] as String?,
      );
}

class OperatingHours {
  final String open;
  final String close;
  const OperatingHours({required this.open, required this.close});

  factory OperatingHours.fromJson(Map<String, dynamic> j) =>
      OperatingHours(open: j['open'] as String, close: j['close'] as String);
}

class Store {
  final String storeId;
  final String ownerUid;
  final String ownerName;
  final String shopName;
  final String phone;
  final String email;
  final String? shopPhotoUrl;
  final String address;
  final StoreLocation location;
  final bool isOpen;
  final bool isActive;
  final bool isVerified;
  final bool isSuspended;
  final int? suspensionEndDate;
  final int strikeCount;
  final int totalStrikes;
  final List<String> availableItemIds;
  final OperatingHours? operatingHours;
  final int totalOrdersAccepted;
  final int totalOrdersFailed;
  final double rating;
  final String? fcmToken;

  const Store({
    required this.storeId,
    required this.ownerUid,
    required this.ownerName,
    required this.shopName,
    required this.phone,
    required this.email,
    this.shopPhotoUrl,
    required this.address,
    required this.location,
    required this.isOpen,
    required this.isActive,
    required this.isVerified,
    required this.isSuspended,
    this.suspensionEndDate,
    required this.strikeCount,
    required this.totalStrikes,
    required this.availableItemIds,
    this.operatingHours,
    required this.totalOrdersAccepted,
    required this.totalOrdersFailed,
    required this.rating,
    this.fcmToken,
  });

  factory Store.fromJson(Map<String, dynamic> j) => Store(
        storeId: j['store_id'] as String,
        ownerUid: (j['owner_uid'] ?? '') as String,
        ownerName: (j['owner_name'] ?? '') as String,
        shopName: (j['shop_name'] ?? '') as String,
        phone: (j['phone'] ?? '') as String,
        email: (j['email'] ?? '') as String,
        shopPhotoUrl: j['shop_photo_url'] as String?,
        address: (j['address'] ?? '') as String,
        location: StoreLocation.fromJson(
          Map<String, dynamic>.from(j['location'] as Map),
        ),
        isOpen: (j['is_open'] ?? false) as bool,
        isActive: (j['is_active'] ?? true) as bool,
        isVerified: (j['is_verified'] ?? false) as bool,
        isSuspended: (j['is_suspended'] ?? false) as bool,
        suspensionEndDate: j['suspension_end_date'] as int?,
        strikeCount: (j['strike_count'] ?? 0) as int,
        totalStrikes: (j['total_strikes'] ?? 0) as int,
        availableItemIds: ((j['available_item_ids'] as List?) ?? const [])
            .map((e) => e.toString())
            .toList(),
        operatingHours: j['operating_hours'] == null
            ? null
            : OperatingHours.fromJson(
                Map<String, dynamic>.from(j['operating_hours'] as Map)),
        totalOrdersAccepted: (j['total_orders_accepted'] ?? 0) as int,
        totalOrdersFailed: (j['total_orders_failed'] ?? 0) as int,
        rating: ((j['rating'] ?? 0) as num).toDouble(),
        fcmToken: j['fcm_token'] as String?,
      );

  Store copyWith({bool? isOpen, List<String>? availableItemIds}) => Store(
        storeId: storeId,
        ownerUid: ownerUid,
        ownerName: ownerName,
        shopName: shopName,
        phone: phone,
        email: email,
        shopPhotoUrl: shopPhotoUrl,
        address: address,
        location: location,
        isOpen: isOpen ?? this.isOpen,
        isActive: isActive,
        isVerified: isVerified,
        isSuspended: isSuspended,
        suspensionEndDate: suspensionEndDate,
        strikeCount: strikeCount,
        totalStrikes: totalStrikes,
        availableItemIds: availableItemIds ?? this.availableItemIds,
        operatingHours: operatingHours,
        totalOrdersAccepted: totalOrdersAccepted,
        totalOrdersFailed: totalOrdersFailed,
        rating: rating,
        fcmToken: fcmToken,
      );
}

class DeliveryBoy {
  final String deliveryBoyId;
  final String storeId;
  final String name;
  final String phone;
  final String? profilePhotoUrl;
  final String googleAccountEmail;
  final bool isActive;
  final String? currentOrderId;

  const DeliveryBoy({
    required this.deliveryBoyId,
    required this.storeId,
    required this.name,
    required this.phone,
    this.profilePhotoUrl,
    required this.googleAccountEmail,
    required this.isActive,
    this.currentOrderId,
  });

  factory DeliveryBoy.fromJson(Map<String, dynamic> j) => DeliveryBoy(
        deliveryBoyId: j['delivery_boy_id'] as String,
        storeId: (j['store_id'] ?? '') as String,
        name: (j['name'] ?? '') as String,
        phone: (j['phone'] ?? '') as String,
        profilePhotoUrl: j['profile_photo_url'] as String?,
        googleAccountEmail: (j['google_account_email'] ?? '') as String,
        isActive: (j['is_active'] ?? true) as bool,
        currentOrderId: j['current_order_id'] as String?,
      );
}

File: backend/models/store.py
from pydantic import BaseModel
from typing import Optional, List


class StoreLocation(BaseModel):
    lat: float
    lng: float
    geohash: Optional[str] = None


class OperatingHours(BaseModel):
    open: str   # "09:00"
    close: str  # "22:00"


class CustomItem(BaseModel):
    custom_item_id: str
    store_id: str
    name: str
    price: float
    unit: str
    image_url: Optional[str] = None
    is_available: bool = True


class DeliveryBoy(BaseModel):
    delivery_boy_id: str
    store_id: str
    name: str
    phone: str
    profile_photo_url: Optional[str] = None
    google_account_email: str
    is_active: bool = True
    current_order_id: Optional[str] = None
    created_at: Optional[int] = None


class Store(BaseModel):
    store_id: str
    owner_uid: str
    owner_name: str
    shop_name: str
    store_type: str = "grocery"   # grocery | fruits | electronics | pharmacy
    phone: str
    email: str
    shop_photo_url: Optional[str] = None
    address: str
    location: StoreLocation
    is_open: bool = False
    is_active: bool = True
    is_verified: bool = False
    is_suspended: bool = False
    suspension_end_date: Optional[int] = None   # epoch ms
    strike_count: int = 0
    total_strikes: int = 0
    available_item_ids: List[str] = []
    operating_hours: Optional[OperatingHours] = None
    self_delivery: bool = False   # owner delivers own orders (shares live GPS) — no delivery partner
    total_orders_accepted: int = 0
    total_orders_failed: int = 0
    onboarded_by: Optional[str] = None          # admin uid
    rating: float = 0.0
    fcm_token: Optional[str] = None
    onboarding_date: Optional[int] = None       # epoch ms
    created_at: Optional[int] = None


class StoreCreateRequest(BaseModel):
    owner_uid: str
    owner_name: str
    shop_name: str
    phone: str
    email: str
    address: str
    lat: float
    lng: float
    operating_hours: Optional[OperatingHours] = None


class StoreSelfRegisterRequest(BaseModel):
    owner_name: str
    shop_name: str
    store_type: str = "grocery"   # grocery | fruits | electronics | pharmacy (mandatory in UI)
    phone: str
    address: str
    lat: float
    lng: float
    operating_hours: Optional[OperatingHours] = None
    self_delivery: bool = False


class StoreProfileUpdateRequest(BaseModel):
    shop_name: Optional[str] = None
    owner_name: Optional[str] = None
    phone: Optional[str] = None
    address: Optional[str] = None
    lat: Optional[float] = None
    lng: Optional[float] = None
    operating_hours: Optional[OperatingHours] = None
    self_delivery: Optional[bool] = None


class StoreToggleRequest(BaseModel):
    is_open: bool


class StoreFCMTokenRequest(BaseModel):
    fcm_token: str
